### PR TITLE
Remove telemetry batch operation over MQTT

### PR DESCRIPTION
**Pull request by abhipsaMisra**

### CODE REVIEW

File: e2e/test/iothub/device/TelemetryE2eTests.cs [hunk 1]
```diff
@@ -22,9 +23,6 @@ public partial class TelemetryE2ETests : E2EMsTestBase
         // The maximum message size for device to cloud messages is 256 KB. We are allowing 1 KB of buffer for message header information etc.
         private const int LargeMessageSizeInBytes = 255 * 1024;
 
-        // The size of a device to cloud message. This exceeds the the maximum message size set by the hub; 256 KB.
-        private const int ExceedAllowedMessageSizeInBytes = 300 * 1024;
```
Comment (abhipsaMisra) [inline e2e/test/iothub/device/TelemetryE2eTests.cs:26]: This was unused.

File: e2e/test/iothub/device/TelemetryE2eTests.cs [hunk 2]
```diff
@@ -36,33 +34,42 @@ public partial class TelemetryE2ETests : E2EMsTestBase
...
         [TestMethod]
         [Timeout(TestTimeoutMilliseconds)]
-        [DataRow(IotHubClientTransportProtocol.Tcp)]
-        [DataRow(IotHubClientTransportProtocol.WebSocket)]
-        public async Task Message_DeviceSendSingleMessage_Amqp(IotHubClientTransportProtocol protocol)
+        [DataRow(TestDeviceType.Sasl, IotHubClientTransportProtocol.Tcp)]
+        [DataRow(TestDeviceType.Sasl, IotHubClientTransportProtocol.WebSocket)]
+        [DataRow(TestDeviceType.X509, IotHubClientTransportProtocol.Tcp)]
+        [DataRow(TestDeviceType.X509, IotHubClientTransportProtocol.WebSocket)]
+        public async Task Message_DeviceSendSingleMessage_Mqtt(TestDeviceType testDeviceType, IotHubClientTransportProtocol protocol)
```
Comment (abhipsaMisra) [inline e2e/test/iothub/device/TelemetryE2eTests.cs:41]: Consolidate X509 and Sas based device tests into the same test group.

**Review by patilsnr — APPROVED** [review]

File: SDK v2 migration guide.md [hunk 3]
```diff
@@ -161,7 +162,7 @@ Find a client you currently use below, read the table of API name changes and us
 | `DeviceClient` | `IotHubDeviceClient` | Specify the service it is a device client for. |
 | `DeviceClient.Dispose()` | `IotHubDeviceClient.DisposeAsync()` | Ensures the client is closed before disposing. |
 | `DeviceClient.SendEventAsync(...)` | `IotHubDeviceClient.SendTelemetryAsync(...)` | Even our public documentation calls this telemetry, so we renamed the method to describe this better.¹ |
-| `DeviceClient.SendEventBatchAsync(...)` | `IotHubDeviceClient.SendTelemetryBatchAsync(...)` | See¹ |
+| `DeviceClient.SendEventBatchAsync(...)` | `IotHubDeviceClient.SendTelemetryBatchAsync(...)` | This is now only supported over AMQP. Support over MQTT has been removed. See¹ |
```
Comment (abhipsaMisra) [inline SDK v2 migration guide.md:165]: Does anyone know what `See¹` refers to? I can't find a reference to this "1" anywhere. Plus, no effect on Ctrl+click.

Comment (drwill-ms) [reply SDK v2 migration guide.md:165]: Line 164 ends in a 1.

Comment (drwill-ms) [inline SDK v2 migration guide.md:165]:
```suggestion
| `DeviceClient.SendEventBatchAsync(...)` | `IotHubDeviceClient.SendTelemetryBatchAsync(...)` | This is now only supported over AMQP. Support over MQTT has been removed. Also, see¹. |
```

File: iothub/device/src/IotHubBaseClient.cs [hunk 4]
```diff
@@ -168,7 +168,7 @@ public async Task SendTelemetryAsync(TelemetryMessage message, CancellationToken
         /// <remarks>
         /// The client instance must be opened already.
         /// <para>
-        /// Use AMQP for a true batch operation. MQTT will just send the messages one after the other.
+        /// This operation is supported only over AMQP.
```
Comment (drwill-ms) [inline iothub/device/src/IotHubBaseClient.cs:171]: How about, this operation is not supported over MQTT and will result in an InvalidOperationException?

Comment (drwill-ms) [reply iothub/device/src/IotHubBaseClient.cs:171]: Also, document exception below.

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Comments inline.

File: iothub/device/src/IotHubBaseClient.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/device/src/IotHubBaseClient.cs:171]: Added.

File: iothub/device/src/IotHubBaseClient.cs [hunk 5]
```diff
@@ -178,6 +181,7 @@ public async Task SendTelemetryAsync(TelemetryMessage message, CancellationToken
         /// <param name="messages">An <see cref="IEnumerable{Message}"/> set of message objects.</param>
         /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
         /// <exception cref="InvalidOperationException">Thrown if the client instance is not opened already.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if this operation is carried out over MQTT.</exception>
```
Comment (drwill-ms) [inline iothub/device/src/IotHubBaseClient.cs:184]:
```suggestion
        /// <exception cref="InvalidOperationException">When this method is called when the client is configured to use MQTT.</exception>
```

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs [hunk 6]
```diff
@@ -419,14 +419,9 @@ public override async Task SendTelemetryAsync(TelemetryMessage message, Cancella
...
-        public override async Task SendTelemetryBatchAsync(IEnumerable<TelemetryMessage> messages, CancellationToken cancellationToken)
+        public override Task SendTelemetryBatchAsync(IEnumerable<TelemetryMessage> messages, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            // Note that this sends all messages at once and then waits for all the acknowledgements. This
-            // is the recommended pattern for sending large numbers of messages over an asynchronous
-            // protocol like MQTT
-            await Task.WhenAll(messages.Select(x => SendTelemetryAsync(x, cancellationToken))).ConfigureAwait(false);
+            throw new InvalidOperationException("Batch telemetry is supported only over AMQP.");
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:424]: I feel like the MQTT transport layer shouldn't know about the other transports. A more cohesive message would be that it is not supported over MQTT. Thoughts?

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs (hunk 6 above)
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:424]: I see your point. I think we'll still need to guide users down the correct path.
How about:
`throw new InvalidOperationException("This operation is not supported over this protocol. Please refer to the doc comments for additional details.");`

**Pull request merged**